nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/pgp/PGPControllerService.java

```diff
@@ -0,0 +1,893 @@
+        problems.add(new ValidationResult.Builder()
+                .subject(CONTROLLER_NAME)
+                .valid(false)
+                .explanation("Controller requires a public key or PBE pass-phrase for Encrypt operations.")
+                .build());
+        return problems;
+    }
+
+
+    /**
+     * This method validates a {@link PropertyContext} for decryption.
+     *
+     * @param context {@link PropertyContext} describing decryption keys
+     * @return validation results, which may be empty if given context is valid
+     */
+    private Collection<ValidationResult> validateForDecrypt(PropertyContext context) {
+        PGPPrivateKey key = getPrivateKey(context);
+        if (key != null) {
+            return Collections.emptySet();
+        }
+
+        char[] passphrase = getPBEPassPhrase(context);
+        if (passphrase != null) {
+            return Collections.emptySet();
+        }
+
+        final List<ValidationResult> problems = new ArrayList<>();
+        problems.add(new ValidationResult.Builder()
+                .subject(CONTROLLER_NAME)
+                .valid(false)
+                .explanation("Controller requires a secret key or PBE pass-phrase for Decrypt operations.")
+                .build());
+        return problems;
+    }
+
+
+    /**
+     * This method validates a {@link PropertyContext} for signing.
+     *
+     * @param context describing signing keys
+     * @return validation results, which may be empty if given context is valid
+     */
+    private Collection<ValidationResult> validateForSign(PropertyContext context) {
+        PGPPrivateKey key = getPrivateKey(context);
+        if (key != null) {
+            return Collections.emptySet();
+        }
+
+        final List<ValidationResult> problems = new ArrayList<>();
+        problems.add(new ValidationResult.Builder()
+                .subject(CONTROLLER_NAME)
+                .valid(false)
+                .explanation("Controller requires a secret key for Sign operations.")
+                .build());
+        return problems;
+    }
+
+
+    /**
+     * This method validates a {@link PropertyContext} for verifying.
+     *
+     * @param context describing verify keys
+     * @return validation results, which may be empty if given context is valid
+     */
+    private Collection<ValidationResult> validateForVerify(PropertyContext context) {
+        PGPPublicKey key = getPublicKey(context);
+        if (key != null && key.isEncryptionKey()) {
+            return Collections.emptySet();
+        }
+
+        final List<ValidationResult> problems = new ArrayList<>();
+        problems.add(new ValidationResult.Builder()
+                .subject(CONTROLLER_NAME)
+                .valid(false)
+                .explanation("Controller requires a public key for Verify operations.")
+                .build());
+        return problems;
+    }
+
+
+    private PGPPublicKey getPublicKey(PropertyContext context) {
```

Comment:
I think additional debug logging should be added to help diagnose potential key loading issues.

Reply:
Added additional debug logging to each of the exception and null-handling cases here and in the private key method.